### ISSUE TITLE
#include cpp file

### ISSUE THREAD

**Issue by LoganDark** [gansm/finalcut#27]
I want to compile this with my application... how do I do that? I can't include the CPP files, nor can I act like `final.h` (define USE_FINAL_H, symlink src/include/final to src/final and include the C++ files in the same order)...

**Comment by gansm** [comment]
Hello LoganDark, without a code example I can't really help you.
Have you ever tried to compile dialog.cpp from Getting Started?

Link: [Getting Started](/gansm/finalcut/wiki/Getting-Started#example-code)

**Comment by LoganDark** [comment]
I've read that already. You don't need a code sample. Read my post.

> I want to compile this with my application

I can't use the header file (i.e. `final.h`) without the rest of the library as well. However, I don't want to use the Makefile and then _link_ my program to the library, I want to compile the library _as part of my program_.

**Comment by gansm** [comment]
If I understand you correctly, you want to link your program against the static libfinal.

```bash
$ g++ -Wall -pedantic -Isrc/include/ dialog.cpp -o dialog src/.libs/libfinal.a -lgpm -ltinfo

$ ldd ./dialog
	linux-vdso.so.1 (0x00007ffde7ffc000)
	libgpm.so.2 => /usr/lib/x86_64-linux-gnu/libgpm.so.2 (0x00007fbdfc3da000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007fbdfc1b0000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fbdfbe2e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fbdfbb2a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fbdfb913000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbdfb574000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fbdfc89d000)
```

**Comment by LoganDark** [comment]
No, as I've said twice already I don't want to link my program with anything. Don't you know what an `#include` is? Online environments like Repl.it don't support linking, only compiling from one entrypoint (== `main.cpp`).

**Comment by gansm** [comment]
You have an environment without a linker?
You want to merge your code with all FINAL CUT files into one big source file using include statements?

**Comment by LoganDark** [comment]
Yes, exactly. Is that possible?

**Comment by gansm** [comment]
I've never tried this before. You can try it and tell me about it. But I'm afraid that the compiler will get problems with the computer resources with such a large amount of code.

**Comment by LoganDark** [comment]
> I'm afraid that the compiler will get problems with the computer resources with such a large amount of code.

I have 16GB of RAM and a quad-core i7, I don't think resources will be a problem.

> You can try it and tell me about it.

I did, in the first post, I needed your help since I couldn't get it working...

---

Anyway, sorry for the long response time, finals week happened. Yay.

I ended up writing my own TUI from scratch because nothing could fit the simple requirement of not having a ridiculously complicated building and linking process. (Even if it was just, say, 2 commands, that's still 2 more than it should be.

Bringing Final Cut in wouldn't have been very efficient, considering I only needed simple cursor movement and input capabilities. I used ANSI escape sequences and a wrapper around this that could detect the arrow keys:

![image](https://user-images.githubusercontent.com/4723091/49729430-50400900-fc2a-11e8-8564-58534278ed32.png)

Leaving this issue open just in case you'd like to make this library work header-only.

**Comment by gansm** [comment]
Even if you have only minor requirements for your programming project, keep in mind that the escape sequences between the terminals may differ.

https://opensource.apple.com/source/emacs/emacs-78.2/emacs/etc/termcap.src

Therefore, I would suggest you at least use a simple termcap implementation to write terminal-independent code.

https://github.com/marcv81/termcap

But there is no character optimization for the terminal like for [S-Lang](https://www.jedsoft.org/slang/), [ncurses](https://www.gnu.org/software/ncurses/) or final cut.

**Comment by LoganDark** [comment]
> keep in mind that the escape sequences between the terminals may differ

[repl.it](https://repl.it) will always use the same terminal. I hardcoded the values for the online environment. It's a toy project for school, not a real application. For a real application running on real systems where I could actually distribute binaries, I'd probably either implement real escape sequence parsing (instead of this hacked-together stuff). repl.it is extremely limited so I had to hack together a non-portable solution.
